File: 2023/Java/app/src/main/java/adventofcode/Day5.java
package adventofcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class Day5 {
    private final long[] seeds;
    private final ArrayList<ArrayList<long[]>> maps = new ArrayList<>();

    public Day5(String data) {
        String[] dataArr = data.split("\n");
        this.seeds = Arrays.stream(
            dataArr[0]
            .split(": ")[1]
            .split(" ")
            ).mapToLong(Long::parseLong)
            .toArray();
        int i = -1;
        for (String line: dataArr) {
            if (line.isEmpty());
            else if (line.charAt(0) >= '0' && line.charAt(0) <= '9') {
                this.maps.get(i).add(Arrays.stream(line.split(" ")).mapToLong(Long::parseLong).toArray());
            } else {
                this.maps.add(new ArrayList<>());
                i++;
            }
        }
    }

    private long locate(long seed, boolean reverse) {
        ArrayList<ArrayList<long[]>> maps = new ArrayList<>();
        for (ArrayList<long[]> map: this.maps) {
            maps.add(new ArrayList<>(map));
        }
        if (reverse) {
            Collections.reverse(maps);
        }
        for (ArrayList<long[]> map: maps) {
            for (long[] submap: map) {
                if (reverse) {
                    submap = new long[]{submap[1], submap[0], submap[2]};
                }
                if (seed >= submap[1] && seed < submap[1] + submap[2]) {
                    seed += submap[0] - submap[1];
                    break;
                }
            }
        }
        return seed;
    }

    private long findRanged(long start, long end, long step) {
        if (step == 0) {
            return end;
        }
        for (long loc = start; loc <= end; loc += step) {
            for (int s = 0; s < this.seeds.length; s += 2) {
                if (locate(loc, true) >= this.seeds[s] && locate(loc, true) < this.seeds[s] + this.seeds[s + 1]) {
                    return findRanged(loc - step, loc, step / 10);
                }
            }
        }
        throw new RuntimeException("No solution found");
    }

    public void run() {
        long single = Integer.MAX_VALUE;
        for (long seed: this.seeds) {
            seed = locate(seed, false);
            single = Math.min(single, seed);
        }
        long ranged = findRanged(0, this.seeds[0] + this.seeds[1], (long) Math.pow(10, 7));
        System.out.println("Part 1: " + single);
        System.out.println("Part 2: " + ranged);
    }
}

File: 2023/Python/day5.py
def parse(data):
    seeds = data[0].split(": ")[1].split()
    seeds = list(map(int, seeds))

    maps = []
    i = -1

    for line in data[2:]:
        if not line:
            pass
        elif line[0].isdigit():
            maps[i].append(list(map(int, line.split())))
        else:
            maps.append([])
            i += 1

    return seeds, maps

def locate(n, maps, reverse=False):
    if reverse:
        maps = maps[::-1]

    for map in maps:
        for submap in map:
            if reverse:
                submap = (submap[1], submap[0], submap[2])
            if n in range(submap[1], submap[1] + submap[2]):
                n += submap[0] - submap[1]
                break
    return n

def find_ranged(start, end, seeds, maps, step):
    if not step:
        return end
    for location in range(start, end + 1, step):
        for s in range(0, len(seeds), 2):
            if locate(location, maps, True) in range(seeds[s], seeds[s] + seeds[s + 1]):
                return find_ranged(location - step, location, seeds, maps, step // 10)

def run(data):
    seeds, maps = parse(data)

    single = float('inf')
    for seed in seeds:
        seed = locate(seed, maps)
        single = min(single, seed)

    span = 0
    for i in range(0, len(seeds), 2):
        span = max(span, seeds[i] + seeds[i + 1])

    '''
    There is a chance that the solution won't be computed
    correctly if your input's direct mapping has narrow
    subranges. If this is the case, try decreasing the step.
    '''

    step = 10**7
    ranged = find_ranged(0, span, seeds, maps, step)

    print(f"Part 1: {single}")
    print(f"Part 2: {ranged}")
